App: Introduce async document recompute.

tritao opened the pull request:
Introduce optional asynchronous recomputation of documents and features to keep the UI responsive during heavy operations. When enabled, recompute requests are processed by background worker thread, allowing the main GUI thread to continue rendering and handling user input.

Errors such as dependency cycles are reported back on the UI thread via callbacks, and the classic synchronous recompute path remains available when the feature is turned off.

This adds the base infrastructure, which will be used by following PRs to enable aborting long running operations.

**Background worker:**

On startup, `Application` spawns a `_recomputeThread` that waits on a `std::condition_variable`.

Requests are enqueued via `queueRecomputeRequest()`, protected by a mutex, and the thread cleanly shuts down in the destructor by signaling `_stopRecomputeThread`.

**Request/Result types:**

`RecomputeRequest` holds pointers to a `Document` or `DocumentObject`, a recursion flag, and a callback.

`RecomputeResult` captures success or exception state.

`isAsyncRecomputeEnabled()` reads this flag to choose async vs. sync.

This enhancement will ensure long recompute operations no longer block the interface, improving user experience while retaining full backwards compatibility, though some more work is needed to fully achieve this.

This work is done as part of an FPA grant: https://github.com/FreeCAD/FPA-grant-proposals/issues/36

tritao committed:
Gui: Add `Enables async document recomputation` preferences option.

ickby commented:
Isn't that too much of a fundamental change to simply start introducing stuff via PR? Is there any documentation on the outcome of the grant proposal? What is the architecture around that feature, how does the proposed synchronization work, what are benefits and drawbacks? Was there any discussion on howto implement it and if that should be done?

furgo16 commented:
I've always considered that all of the above can be done in a PR:
- Submitting a PR does not imply it will be merged.
- Its description can provide info on expected outcome, architecture, implementation... anything that can be put in a document.
- Anyone can review it and provide their feedback in the comments
- Submitting initial code gives the ability to discuss an actual implementation, comment the code line by line if necessary, provide alternative branches, etc.

ickby commented:
I would usually agree for normal features. But the proposed functionality is going to the very core of freecad and will touch most likely almost all of its fundamental systems, everywhere in the codebase. This will be huge, the code alone can impossibly be part of a single PR. This has to be judged on architecture level before even starting thinking about code. 

At least mark the PR as draft. And inform us when a architecture review can be performed (not code, architecture).

ickby commented:
> I don't expect it to be all that huge

Ok, seems we have vastly different expectations of what is needed to make this work. But I'm always happy to be proven wrong, so lets go into some details. The current code of this PR moves the recompute calls to a new thread and keeps the main thread running seperatly. Here are my first basic questions for that architecture:

1. I don't see anything for data synchronization or access protection. How do you plan to avoid data races and race conditions on the document data and the application state between recompute and main thread?
2. I'm unsure how you want to handle thread safety of freecads current code. The recompute in the new thread will call functions all over the place, and as currently proposed the main thread can call all of them too. Do you plan to make all of FreeCAD functions reentrant?
3. There is nothing in place to protect both threads using the same external libraries at the same time. This requires by definition that those libraries are thread safe. How do you plan to go around that, or do plan to introduce thread safety as a requirement for FreeCAD dependencies?

ickby commented:
Thanks for the explanation, that gives a better picture. I think/hope I understand your design correctly.

>The proposal is that the GUI will block any operations that allow mutation of data while async recompure requests are going on.

IMHO that proposal is flawed, the GUI and direct user operations may be the main source of data access, but not the only one. The full QT event loop on the main thread is running in parallel to the recompute thread, and it can trigger any kind of code independent of the user, mostly due to delayed operations. Things on top of my head:

 - QTimer based events
 - QProcess and QThread based code
 - QNetworkAccessManager based web requests
 - A python asyncio loop instance based on the qt event loop
 
FreeCAD modules can use any of those Qt/PySide provided features, and they should be able to without creating random errors.

Additionally problematic, but most likely a easier to solve:
- Macros/Console code executed from within FreeCAD that have simply additional code after a recompute call
- Python code passed to FreeCADCMD executable

How do you intend to prevent data access or parallel execution of thread-unsafe code by these methods?

ickby commented:
> Python code is already serialized by the GIL, so I don't expect it to be much of an issue.
I disagree. The GIL ensures non-parallel execution of python code only, hence it ensures that no python code of both threads run at the same time.  To my understanding there can be python code owning the GIL on the main thread in parallel to c++ code running on recompute thread, both using the exact same data and functions at the same time. Together with  the async nature of the QT event loop this is the big risk python modules face.

> Also I don't think it's reasonable to try and preemptively detect or support any kind of unsafe patterns that don't play well an async world in the rest of the ecosystem. If such patterns occur, the addons should be modified

IMHO this cannot be the mindset here. We cannot just hope and pray that external developers do not use certain provided FreeCAD core API functionality like the qt event loop, if the consequences of using it are crashes of the application. I'm perfectly fine if things give errors in a controlled and deterministic way, and developers adopting to that and integrate patterns to make it work. But the way it is now everyone not using the correct patters will introduce almost random errors and crashes based on a million different possible states of the application and user interaction, which is not debugable. Additionally, safe patterns are not possible in the currently proposed code, there is nothing provided for it, currently the only workaround is to not use this functionality at all and hence breaking many use cases of FreeCAD.

There has to be a clear plan how to handle the python API for this feature, what are the patterns developers need to adapt to, and how to provide application safety and controlled failing/errors for the case the developers do not use the correct patterns. We have to be robust for external addons that are out of review control of the core devs.

> Overall I don't think any of those issues impact the core architecture proposed in this PR, with the worker thread and recompute queue. All that can be implemented on top of the system here, and I don't think impacts its design? Let me know if you think otherwise.
Yes, i think otherwise. The code in this PR on itself is highly broken. I know it is deactivated, but still, why merge broken code that is completely unsafe just with the hope that it will work later, when it is not even shown yet that it will?

My personal opinion: For such a fundamental high risk change one cannot simply wing it, the following things need to happen before merging anything:
1. Provide a fully running, safe and thought out implementation in a branch for everyone to test and look at. It must be shown to work before merging.
2. Provide a detailed list of all benefits and drawbacks of that implementation to the developers and community, incl. new maintenance needs, new restrictions, incompatibilities, required patterns, impact on test procedures etc. 
3. Let the dev. community decide if the benefits are worth the drawbacks. I know that it was funded already, but the initial proposal did not include any of the now found drawbacks like breaking some python modules. It is only fair for everyone to get to decide if this is wanted when all relevant information is given. This needs to be maintained by every developer for forever, it must be clear what this means before committing to it.

My 2 cents. If the rest of the dev. community thinks differently than please go ahead.

ickby commented:
> Are you thinking specifically of issues with multiple parallel Qt event loops?
Like an addon creating a new QThread-based event loop and calling into native FreeCAD APIs?

No, I'm thinking of asynchronously fired events from the event loop that trigger code execution. Imagine the following scenarios:
1. Running timing based periodic events, for example to have animations of assemblies and models ([example workbench](https://github.com/QuentinTournier40/AnimationFreeCAD))
2. Having running something in a QProcess that will finished at unknown time, but triggers a event when done. For example using a external tool to do some processing and then load the result ([example workbench](https://github.com/FreeCAD/FreeCAD-render))
3. Having a server running in FreeCAD, or using FreeCAD as client in a server connection: There may be requests incomming at any time or request finished with arbitrary delay, and trigger code ([example workbench](https://github.com/bonninr/freecad_mcp))

Right now all these scenarios have solid guarantees about the application state when their triggered code is running. It is safe to run any python code with these techniques, as it should be. But to what I understand right now from the proposal these guarantees will not hold in the future, as the relevant events can fire parallel to the recompute thread. Therefore all these scenarios have to somehow check now what is safe and what not, with potentially crashing FreeCAD if they get it wrong.

Again, I'm fine with some patterns these scenarios need to adopt to to work properly. But it still needs to be ensured that not following those pattern does not potentially crash the application, but leads to python error for the modules when doing unsafe calls.

ickby commented on src/App/DocumentObserverPython.h:
```diff
@@ -83,8 +83,6 @@ class AppExport DocumentObserverPython
     /** Called when a given object is recomputed */
     void slotRecomputedObject(const App::DocumentObject& Obj);
-    /** Called before an observed document is recomputed */
-    void slotBeforeRecomputeDocument(const App::Document& Doc);
```
Experimental code should not break the python API by removing features.

ickby replied:
There are many addons outside the freecad core using the python API, all over github and potentially even not made public but used for private / restricted use cases. This has to be considered.

ickby replied:
Assembly3 uses it and will be broken without: https://github.com/realthunder/FreeCAD_assembly3

> There could be some private addon using it, but this signal is incompatible with an async world.

But than there has to be a deliberate decision to break the python API and removing functionality. I don't like doing this just on the fly during development of a maybe feature without discussion or notification to anyone, no deprecation messages, nothing... Perfectly fine for a experimental branch, but not for merging into main.

ickby replied:
thanks!

luzpaz commented:
> There is another debate here beyond the topic. I think it's a valuable fundamental argument. Can experimental features be included in the core if they are sufficiently isolated? How should this isolation be? I couldn't find a similar example.

This happens often. To enable them the user needs to manipulate/'fine tune' via the Parameter Editor. See https://wiki.freecad.org/Fine-tuning

maxwxyz commented:
@tritao I will test this PR. Is there any need for specific tests?
Could you resolve conflicts?

maxwxyz reviewed (approved):
I'm not sure how to test this.
Playing around it seems to have no negative side effects. I only played with PartDesign stuff, no other workbenches.
Patterns seem to be faster, but creating a polar pattern with the PartDesign Example (last pocket feature) of 50 freezes the app, there is no feedback or anything:
<img width="3316" height="1416" alt="grafik" src="https://github.com/user-attachments/assets/a06697e4-6951-4b02-b212-a7dde843185f" />

I'm not sure if this is related to @kadet1090 previews as disabling them will increase speed significantly and the app does not freeze without previews.

Edit: Seems to work now after restarting. It is a lot faster. Not sure if previews and the draggers @captain0xff could benefit from that too? Or other WBs, like dragging in assembly or sketcher? IIRC it calculates the feature the sketch is used in when changing the sketch with each change or drag @AjinkyaDahale @PaddleStroke 

IMO if merged, this should be on by default (or even without a preference).

kadet1090 commented on src/App/Application.cpp:
```diff
@@ -264,10 +264,21 @@ Application::Application(std::map<std::string,std::string> &mConfig)
 }
 
-Application::~Application() = default;
+Application::~Application() {
```
this should follow the style of he file in general.

kadet1090 commented on src/App/MainThreadSignal.h:
```diff
@@ -0,0 +1,226 @@
+    static inline void invoke(std::function<void()>&& fn, bool blocking) {
+        auto* f = invoke_slot();
+        if (f) f(std::move(fn), blocking);
+        else   fn(); // no hooks, run inline
```
That is not our code style. Why isn't it clang formatted?

maxwxyz commented:
> I am curious what happens in the case when some parameter that dictates the dragger orientation or position is changed but not computed yet. Like say checking the reverse checkbox for a pad operation that takes lots of time. 

@captain0xff  do not see an improvement with the draggers or reversing the helix. Neither with or without previews active. See yourself:

https://github.com/user-attachments/assets/31e29fab-ca40-4e08-bfcc-0aeba01d4e25

furgo16 commented:
I created a build of this branch, for those who want to test it without having to compile it: https://github.com/furgo16/FreeCAD/releases/tag/preview-async-document-recompute

pieterhijma commented on src/App/Application.cpp:
```diff
@@ -639,6 +650,20 @@ bool Application::isClosingAll() const {
+        std::lock_guard<std::mutex> lock(_recomputeMutex);
+        _recomputeRequests.push_back(req);
+    }
+    _recomputeCV.notify_one();
```
I would prefer to make use of the constructs you define yourself. This would have been more clear for me.

```suggestion
    notifyRecomputeWorker();
```

pieterhijma reviewed (changes requested):
Well done! I'm looking forward to turn on the async recompute. For now I have some comments. I will also test the feature later and will review once more.

maxwxyz commented:
@tritao ping for feedback and some conflicts

tritao committed:
App: Introduce async document recompute.

Introduce optional asynchronous recomputation of documents and features
so the GUI can stay responsive during heavier operations. When enabled,
recompute work is queued for a background worker while completion and
failure handling stays on the main thread.

RecomputeRequest stores stable document and object identifiers plus the
completion callback, so queued work can resolve its targets safely even
if document lifetime changes while the request is waiting in the queue.
Application tracks queued and in-progress documents under a mutex and
uses a separate atomic stop flag so shutdown can wake the worker without
first taking the queue lock.

The classic synchronous recompute path remains available when async
recompute is disabled. This adds the base infrastructure used by the
follow-up GUI and test commits in this branch.

This work is done as part of an FPA grant:
https://github.com/FreeCAD/FPA-grant-proposals/issues/36

tritao committed:
App: Introduce main thread signals and switch document signals to use it.

tritao committed:
Gui: Introduce async recompute support for `StdCmdRefresh` command.

tritao committed:
Gui: Fail fast on worker-thread GUI access

tritao committed:
Gui: Guard Python GUI entry points on main thread

tritao committed:
Gui: Guard command entry points on main thread

tritao committed:
Test: Cover async recompute thread guards

tritao committed:
Doc: describe async recompute threading rules

tritao committed:
Gui: Cleanup unused Qt includes.

tritao committed:
Gui: enable async recompute by default

tritao committed:
Gui: Harden autosave around unstable document states

Block autosave while a document is in unstable App or Gui state and
retry the pending save once the document becomes stable again.

What this changes:
- adds App::Document::signalBecameStable and documents the recompute-to-stable transition
- derives App-side recovery snapshot eligibility from transaction state instead of mirroring it in DocumentP
- extracts recovery snapshot writing into App::writeRecoverySnapshotToTransientDir() in src/App/RecoverySnapshot.*
- adds FreeCAD.writeRecoverySnapshotToTransientDir(...) and aligns the App-side Python API with the new App entry point
- teaches AutoSaver to coalesce pending work with consumePendingAutosave(), restorePendingAutosave(), and schedulePendingAutosaveRetry()
- keeps autosave blocked while a document is PartialDoc, TempDoc, Recomputing, in an App transaction, or in a Gui transaction

Net effect:
Autosave no longer attempts recovery writes from transient states,
skipped writes are retried once the document becomes stable again, and
the recovery snapshot entry point now lives in App instead of Document.

maxwxyz reviewed (approved):
Tested again. I did not find any issues.

tritao commented:
Updated with a fix for Windows-specific build failure.

pieterhijma commented on src/App/Application.cpp:
```diff
@@ -206,18 +207,129 @@ using namespace App;
+    return request;
+}
+
+bool requestTargetsDocument(const RecomputeRequest& request, const std::string& documentName)
```
Now that the document name is registered, is there a chance that the document name is changed while requests are queued? Would it be necessary to listen to `signalRenameDocument`?

tritao replied:
Good point. I documented the invariant, queued requests use the internal document name, which is assigned at creation time and does not change afterward.

pieterhijma replied:
Great, saw it.

pieterhijma commented on src/App/Application.cpp:
```diff
@@ -206,18 +207,129 @@ using namespace App;
+RecomputeRequest takeNextRecomputeRequest(std::vector<RecomputeRequest>& requests)
+{
+    RecomputeRequest request = std::move(requests.front());
+    requests.erase(requests.begin());
```
Is there a situation in which requests has many items? For example, if many objects are selected and recomputed? Vector erase is O(n).

tritao replied:
I looked at changing this to `std::deque`, but left it as `std::vector` for now. The queue should stay small in practice, I don't think the queue will grow big at all and preferred to keep this simple unless we have evidence of a bottleneck.

pieterhijma replied:
Ok, totally agree.

tritao replied:
> Ok, totally agree.

I've ended up just making the switch on the latest version, by the way.

pieterhijma replied:
Ah ok, because it may be larger than you anticipated or another reason?

pieterhijma commented on src/App/Application.cpp:
```diff
@@ -206,18 +207,129 @@ using namespace App;
+    const auto recomputeObjects = Document::getDependencyList(recomputeRoots, Document::DepSort);
+
+    return std::all_of(recomputeObjects.begin(),
+                       recomputeObjects.end(),
```
```suggestion
    return std::ranges::all_of(recomputeObjects,
```

tritao replied:
Applied.

pieterhijma commented on src/App/Application.cpp:
```diff
@@ -647,34 +783,109 @@ std::string Application::getUniqueDocumentName(const char* Name, bool tempDoc) c
+
+void Application::queueRecomputeRequest(RecomputeRequest req)
+{
+    if (App::MainThreadSignalConfig::hasHooks() && !canRecomputeRequestOnWorker(req)) {
```
I don't fully understand this condition. It leaves the opening that even if a request cannot be computed on the worker, it is still scheduled on the worker (when `hasHooks()` is false and `!canRecomputeRequestOnWorker(req)` is true.

Could you please elaborate on that? If this is indeed the right condition, I would appreciate a comment explaining it.

tritao replied:
Good catch. I changed the logic so `canRecomputeRequestOnWorker()`  is now the decisive check: if a request is not worker-safe, it never goes onto the recompute worker.

pieterhijma commented on src/App/Application.cpp:
```diff
@@ -647,34 +783,109 @@ std::string Application::getUniqueDocumentName(const char* Name, bool tempDoc) c
+    std::unique_lock<std::mutex> lock(_recomputeMutex);
+    _recomputeStateChanged.wait(lock, [this, &documentName] {
+        return _recomputeDocumentsInProgress.find(documentName)
+            == _recomputeDocumentsInProgress.end();
```
```suggestion
        return _recomputeDocumentsInProgress.contains(documentName);
```

tritao replied:
Applied.

pieterhijma commented on src/App/Application.cpp:
```diff
@@ -647,34 +783,109 @@ std::string Application::getUniqueDocumentName(const char* Name, bool tempDoc) c
+
+    auto newEnd = std::remove_if(
+        _recomputeRequests.begin(),
+        _recomputeRequests.end(),
```
```suggestion
    auto newEnd = std::ranges::remove_if(
        _recomputeRequests,
```

tritao replied:
Applied a variation.

pieterhijma commented on src/App/Application.h:
```diff
@@ -993,6 +1056,27 @@ class AppExport Application
 
+    // Worker thread for processing pending recompute requests.
+    std::thread _recomputeThread;
+    std::mutex _recomputeMutex;
```
Often when mutexes are introduced, they govern a set of variables/members. That is the case here as well. To make the code maintainable for others as well, I would find it very helpful if the mutex and the variables it governs are declared close together and that there is an explicit and very clear comment that this particular mutex protects these variables. Could you please add that?

tritao replied:
Agreed. I added a comment documenting that `_recomputeMutex` protects `_recomputeRequests` and `_recomputeDocumentsInProgress`.

pieterhijma replied:
I would still prefer that `_recomputeMutex` and `_recomputeDocumentsInProgress` are declared directly beneath it as if it defines some sort of scope or so, messaging the reader: "This is what belongs together and should stay belonging together":
```suggestion
    // //////////
    // Protects ...
    std::mutex _recomputeMutex;
    std::deque<RecomputeRequest> _recomputeRequests;
    std::set<std::string> _recomputeDocumentsInProgress;
    // //////////
```

In my experience, the fact that a mutex governs more than one variable often means that there is a strong logical connection between the two variables. Often, in a later stage, others that are maintaining the code miss that point and create "disconnects", for example modifying one of these variables without having the lock or introducing a bit more state that should be governed by the lock as well without realizing this strong connection between everything.

pieterhijma commented on src/App/Application.cpp:
```diff
@@ -647,34 +783,109 @@ std::string Application::getUniqueDocumentName(const char* Name, bool tempDoc) c
+            return requestTargetsDocument(request, documentName);
+        }
+    );
+    _recomputeRequests.erase(newEnd, _recomputeRequests.end());
```
O(n) again for erase. Perhaps an issue if recomputeRequests can contain many elements.

tritao replied:
Same reasoning as the earlier comment.

pieterhijma commented on src/App/Application.h:
For me this file has weird line endings. Could it be mixed DOS or Unix or so?

tritao replied:
Yes, unfortunately `Gui` and `App` are full of mixed line endings file, I've tried "fixing" this so its cleaner.

pieterhijma commented on src/Gui/CommandDoc.cpp:
```diff
@@ -1686,35 +1688,98 @@ StdCmdRefresh::StdCmdRefresh()
+
+    // If the user wants to proceed, enqueue another recompute request without
+    // the cycle-check option so the document recomputes like the legacy path.
+    App::RecomputeRequest newRequest = App::RecomputeRequest::fromDocument(*document, true);
```
```suggestion
    App::RecomputeRequest newRequest = App::RecomputeRequest::fromDocument(*document, /*force=*/true);
```

tritao replied:
Applied.

pieterhijma commented on src/Gui/AutoSaver.h:
```diff
@@ -45,19 +48,33 @@ class ViewProvider;
     Connection documentMod;
+    Connection documentStable;
+    std::string documentName;
+    mutable std::mutex touchedMutex;
```
Also for this mutex, I would appreciate it if you could make very clear which data exactly it protects, because it is not only the touched set.

tritao replied:
Agreed. I renamed the mutex to `autosaveStateMutex` and documented exactly which members it protects

pieterhijma commented on src/Gui/AutoSaver.cpp:
```diff
@@ -48,6 +51,93 @@ FC_LOG_LEVEL_INIT("App", true, true)
+public:
+    explicit AutoSaveSnapshotGuard(AutoSaveProperty& saver)
+        : saver(saver)
+        , touchedSnapshot(saver.takeTouchedSnapshot(dirtyGenerationSnapshot))
```
This could benefit from extra documentation explaining the overall logic. I was first mystified on where dirtyGenerationSnapshot was set, only figuring out it was set in `takeTouchedSnapshot()`. It makes sense now, but this was one of the first things I struggled to understand.

tritao replied:
I expanded the documentation here and also simplified the code so the snapshot lifecycle is more explicit.

pieterhijma commented on src/Gui/AutoSaver.cpp:
```diff
@@ -48,6 +51,93 @@ FC_LOG_LEVEL_INIT("App", true, true)
+
+    bool commit()
+    {
+        bool needsFlush = saver.finishSuccessfulSave(dirtyGenerationSnapshot);
```
I find this logic difficult to understand. `finishSuccessfulSave()` returns a boolean stating if an extra flush is required given the dirty generation of the snapshot. It doesn't really finish something, but I get from the context that if you do a commit, then a successful save must have occurred. It's just not clear if additional saves are required. If so then we set `pendingAutosave` to true. Is that right?

Perhaps we have to look into terminology as well. New terms:
- document stability: a document is not stable if there's an asynchronous recompute
- mark pending autosave: if a document is not stable, an autosave is marked to be pending
- touched snapshots: we take snapshots of the touched state associated with a dirty generation
- flushing pending autosaves: autosaves that are pending are flushed, written to the autosave file.
- queueing to flush pending autosaves: When the document becomes stable, a flush for pending autosaves is queued. This changes state, namely flush scheduled.
- flush scheduled state: so queueing a flush changes the flush scheduled state

Even though a flush is scheduled, it is apparently possible that there is no pending autosave.

In the naming of this line, it is clear that concurrent constructs such as mutexes don't compose well. You want to check and set things. I would adopt the "and" convention to make atomicity explicit, in form example `compareAndSwap`, `testAndSet`.

```suggestion
        bool requiresFlush = saver.clearAndTestFlushRequired(dirtyGenerationSnapshot);
```

Perhaps you have a better way to rename it.

It would be great for maintainability of the code if you could add documentation like I'm providing in this comment, because it is very challenging to get this from just reading the code.

tritao replied:
Agreed. I renamed this to `commitAndTestFlushRequired()` so the mutate + test behavior is explicit.

pieterhijma commented on src/Gui/AutoSaver.cpp:
```diff
@@ -89,6 +179,26 @@ void AutoSaver::renameFile(QString dirName, QString file, QString tmpFile)
+        return;
+    }
+
+    if (!it->second->beginPendingAutosaveFlush()) {
```
This name is also difficult to understand for me. It has the same problem that it does a set and test. I would suggest a similar name:

```suggestion
    if (!it->second->clearAndTestPendingAutosave()) {
```

Again, you may find a better name.

tritao replied:
Agreed. I renamed this to `clearFlushScheduledAndTestPendingAutosave()`.

pieterhijma commented on src/Gui/AutoSaver.cpp:
```diff
@@ -89,6 +179,26 @@ void AutoSaver::renameFile(QString dirName, QString file, QString tmpFile)
+    }
+
+    try {
+        saveDocument(it->first, *it->second);
```
This looks like a form of recursion (via the event loop) and makes it very difficult for me to reason about. The recursion is:

```
saveDocument()
queuePendingAutoSaveFlush()
flushPendingSave() (via a singleshot)
saveDocument()
```

Can this lead to problems in any way?

tritao replied:
I don't think so.

It is re-entry via the event loop, but not direct stack recursion.

  `saveDocument()` does not call `flushPendingSave()` inline. `queuePendingAutosaveFlush()` posts a `QTimer::singleShot(0, ...)`, so the current save pass returns first and the retry runs later on the next GUI event-loop turn.

So this does not cause unbounded stack growth.

The remaining concerns are multiple queued retries, retries after document deletion, and repeated retries while the document keeps changing. The current code handles those as follows:
- `flushScheduled` ensures there is at most one queued retry per document
- `flushPendingSave()` re-checks that the document still exists before saving
- repeated retries are still possible if the document keeps changing, but only one queued retry exists at a time

So this can lead to repeated deferred save attempts, but not recursive stack growth.

pieterhijma replied:
Yes, I see it.

pieterhijma commented on src/Gui/AutoSaver.cpp:
```diff
@@ -200,39 +346,17 @@ void AutoSaver::saveDocument(const std::string& name, AutoSaveProperty& saver)
+        "Save auto-recovery file in %fs\n",
+        Base::TimeElapsed::diffTimeF(startTime, Base::TimeElapsed())
+    );
+    if (saveSnapshot.commit()) {
```
The "test-and-set" paradigm moves here as well then. I would name it perhaps:

```suggestion
    if (saveSnapshot.commitAndTestFlushRequired()) {
```

This would have made it easier for me to understand.

tritao replied:
Applied.

pieterhijma commented on src/Gui/AutoSaver.cpp:
```diff
@@ -264,13 +387,87 @@ AutoSaveProperty::AutoSaveProperty(const App::Document* doc)
+
+    pendingAutosave = false;
+    return false;
+}
```
This could perhaps be expressed in a better way. I particularly don't like the "true;true" and "false;false" pattern, perhaps:
```suggestion
    return (pendingAutosave = (dirtyGeneration != dirtyGenerationSnapshot))
}
```

tritao replied:
Applied.

pieterhijma commented on src/Gui/AutoSaver.cpp:
In general, the new logic in this file is difficult to understand. It would be great if more documentation would be added. I've already made a small start in a comment.

tritao replied:
Agreed, I added a state-model comment and simplified the implementation a bit with `SaveSnapshot`.

pieterhijma commented on src/Mod/Fem/Gui/TaskPostBoxes.cpp:
```diff
@@ -219,14 +219,17 @@ TaskPostWidget::TaskPostWidget(
-        )
-    );
+    if (auto object = getObject()) {
+        if (auto document = object->getDocument()) {
```
Perhaps better to not nest these checks, happy path without branches?

tritao replied:
Agreed.

pieterhijma reviewed (changes requested):
Looks good in general! This is a very thorough review checking if I could maintain the code myself. In particular I have now questions on the logic in AutoSaver.

Since this is possibly the first time that concurrency is used in this way, I would also prefer to set the example in a couple of conventions dealing with concurrent code, documentation of which variables are protected and adopting "test-and-set"-like naming.

I'm sorry for the many comments. It's a large PR now, but I hope you can still appreciate the comments.

tritao commented:
> I'm sorry for the many comments. It's a large PR now, but I hope you can still appreciate the comments.

Thanks so much for your review, your comments very much appreciated and have led to good improvements.

I've updated the PR addressing them, let me know if you are happy with its current state or would like additional changes.

tritao committed:
Test: cover recovery snapshots and autosaver flushes

tritao committed:
Test: add opt-in autosaver stress driver

tritao commented:
Updated, latest round fixes some minor issues, and adds some testing around the new auto-saver code.

pieterhijma commented on src/Gui/AutoSaver.cpp:
```diff
@@ -138,101 +148,58 @@ void AutoSaver::slotDeleteDocument(const App::Document& Doc)
 }
 
-void AutoSaver::saveDocument(const std::string& name, AutoSaveProperty& saver)
+bool AutoSaver::saveDocument(const std::string& name, AutoSaveProperty& saver)
```
I don't think the boolean is used anywhere. If I'm wrong, please document what the return value means.

```suggestion
void AutoSaver::saveDocument(const std::string& name, AutoSaveProperty& saver)
```

pieterhijma commented on src/Gui/AutoSaver.cpp:
```diff
@@ -257,186 +223,96 @@ void AutoSaver::timerEvent(QTimerEvent* event)
-RecoveryWriter::~RecoveryWriter() = default;
-
-bool RecoveryWriter::shouldWrite(const std::string& name, const Base::Persistence* object) const
+bool AutoSaveProperty::beginSavePass()
```
This name is bit unintuitive name for what it does. Can also be refactored a bit as `flushScheduled` is set to `false` in both branches.

pieterhijma commented on src/Gui/AutoSaver.h:
```diff
@@ -26,38 +26,65 @@
 {
-class ViewProvider;
-
+/**
```
Nice, thanks.

tritao commented:
Rebased to fix conflicts.

pieterhijma commented on src/App/Application.h:
```diff
@@ -993,6 +1059,28 @@ class AppExport Application
+    std::mutex _recomputeMutex;
+    std::condition_variable _recomputeRequestAvailable;
+    std::condition_variable _recomputeStateChanged;
+    std::atomic<bool> _stopRecomputeThread{false};
```
In this particular case, this bool does not need to be governed by the lock. It would be very helpful to point that out, why exactly, and why it needs to be atomic.

pieterhijma commented on src/App/Document.cpp:
```diff
@@ -336,6 +450,7 @@ std::vector<std::string> Document::getAvailableRedoNames() const
 int Document::openTransaction(TransactionName name, int tid) // NOLINT
 {
     if (tid != NullTransaction && tid == d->bookedTransaction) {
+        d->recoveryWriteBlocked = true;
```
I'm not so happy with this extra piece of state that can very easily be misaligned with what it is preventing if someone else introduces changes to the code. For example if work is done on transactions by another contributor, it could be the case that a condition is created in which `recoverWriteBlocked` should also be true.

Also a drawback is that in this particular function I'm wondering what the correct value of `recoveryWriteBlocked` is. There are multiple solutions: is the value of `recoveryWriteBlocked` implicitly false? Or should the value of `recoveryWriteBlocked` be what it was before? From this function I can't easily judge which is the case.

Would it be possible to define a function that can check at the moment you need this value if a recovery write should be blocked? For example `thereShouldNotBeAnythingThatResemblesAnOpenTransaction()` or so.

The state also seems to be closely aligned with `bookedTransaction`.

Well, I don't have a definitive solution, but it would be great if we could reduce the amount of state that we need to track.

pieterhijma commented on src/App/Document.cpp:
```diff
@@ -1872,6 +2017,38 @@ bool Document::saveCopy(const char* file) const
+        && !isPerformingTransaction();
+}
+
+bool Document::writeRecoverySnapshotToTransientDir(const RecoverySaveOptions& options) const
```
To me having this in App seems a better choice in general, but I'm wondering if it is just a refactor that makes sense or if you really needed it to be here, for example for the `canWriteRecoverySnapshot()` and the new boolean state you introduced.

tritao replied:
I agree, I will refactor this.

pieterhijma replied:
No, I don't think a refactor for this is needed! You moved this stuff from Gui to App. I'm just wondering why you did it. Or is your comment a reaction to the boolean state above?

tritao replied:
It was originally in the auto saving code, but I decided to reduce the coupling by extracting it out to `Document`, though I think it left too much app-level code inside it, as you pointed out. The latest iteration fixes it by splitting the relevant code between Document/App.

pieterhijma reviewed (commented):
Thanks for the changes. Looks better, but I still have some comments. I need to verify a couple of more things, so I will do another review, but I wanted to submit this feedback right now.

pieterhijma commented on src/Mod/Test/AutoSaverStress.py:
Unfortunately, I can't run the stress test well because I'm running into this [this issue](https://github.com/FreeCAD/FreeCAD/issues/28307).

<details>
<summary>stacktrace</summary>
<pre>
#1  0x00007fffee8a7af3 in __pthread_kill_internal (threadid=<optimized out>, signo=6) at pthread_kill.c:89
89	  return __pthread_kill_implementation (threadid, signo, 0);
(gdb) f 2
Downloading 1.07 K source file /usr/src/debug/glibc/glibc/signal/../sysdeps/posix/raise.c...
#2  0x00007fffee84d1a0 in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
26	  int ret = __pthread_kill (__pthread_self (), sig);
(gdb) f 3
Downloading 3.07 K source file /usr/src/debug/glibc/glibc/stdlib/abort.c...
#3  0x00007fffee8345fe in __GI_abort () at abort.c:77
77	  raise (SIGABRT);
(gdb) f 4
#4  0x00007ffff5076236 in messageHandler (type=QtFatalMsg, context=..., msg=...) at /home/pieter/current/fpa/freecad-pr-21292/src/Gui/Application.cpp:2329
2329	            abort();  // deliberately core dump
(gdb) f 5
#5  0x00007fffeeef8e58 in ?? () from /usr/lib/libQt6Core.so.6
(gdb) f 6
#6  0x00007fffeeeed3d6 in ?? () from /usr/lib/libQt6Core.so.6
(gdb) bt
#0  __pthread_kill_implementation (threadid=<optimized out>, signo=signo@entry=6, no_tid=no_tid@entry=0) at pthread_kill.c:44
#1  0x00007fffee8a7af3 in __pthread_kill_internal (threadid=<optimized out>, signo=6) at pthread_kill.c:89
#2  0x00007fffee84d1a0 in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
#3  0x00007fffee8345fe in __GI_abort () at abort.c:77
#4  0x00007ffff5076236 in messageHandler (type=QtFatalMsg, context=..., msg=...) at /home/pieter/current/fpa/freecad-pr-21292/src/Gui/Application.cpp:2329
#5  0x00007fffeeef8e58 in ?? () from /usr/lib/libQt6Core.so.6
#6  0x00007fffeeeed3d6 in ?? () from /usr/lib/libQt6Core.so.6
#7  0x00007fffeee9417e in QMessageLogger::fatal(char const*, ...) const () from /usr/lib/libQt6Core.so.6
#8  0x00007fffeee91848 in qt_assert_x(char const*, char const*, char const*, int) () from /usr/lib/libQt6Core.so.6
#9  0x00007ffff5581dce in QtPrivate::assertObjectType<Gui::MainWindow> (o=0x7fffffffcc00) at /usr/include/qt6/QtCore/qobjectdefs_impl.h:105
#10 0x00007ffff5c6a4f1 in QtPrivate::FunctorCall<std::integer_sequence<unsigned long, 0ul>, QtPrivate::List<QMdiSubWindow*>, void, void (Gui::MainWindow::*)(QMdiSubWindow*)>::call (f=(void (Gui::MainWindow::*)(Gui::MainWindow * const, QMdiSubWindow *)) 0x7ffff5c2daa8 <Gui::MainWindow::onWindowActivated(QMdiSubWindow*)>, o=0x7fffffffcc00, arg=0x7fffffffbe70) at /usr/include/qt6/QtCore/qobjectdefs_impl.h:125
#11 0x00007ffff5c67264 in QtPrivate::FunctionPointer<void (Gui::MainWindow::*)(QMdiSubWindow*)>::call<QtPrivate::List<QMdiSubWindow*>, void> (f=(void (Gui::MainWindow::*)(Gui::MainWindow * const, QMdiSubWindow *)) 0x7ffff5c2daa8 <Gui::MainWindow::onWindowActivated(QMdiSubWindow*)>, o=0x7fffffffcc00, arg=0x7fffffffbe70) at /usr/include/qt6/QtCore/qobjectdefs_impl.h:174
#12 0x00007ffff5c64425 in QtPrivate::QCallableObject<void (Gui::MainWindow::*)(QMdiSubWindow*), QtPrivate::List<QMdiSubWindow*>, void>::impl (which=1, this_=0x555555a31080, r=0x7fffffffcc00, a=0x7fffffffbe70, ret=0x0) at /usr/include/qt6/QtCore/qobjectdefs_impl.h:545
#13 0x00007fffeefd8f0f in ?? () from /usr/lib/libQt6Core.so.6
#14 0x00007ffff02f5c5b in ?? () from /usr/lib/libQt6Widgets.so.6
#15 0x00007fffeefd92da in ?? () from /usr/lib/libQt6Core.so.6
#16 0x00007ffff02fedcb in ?? () from /usr/lib/libQt6Widgets.so.6
#17 0x00007ffff02f42d5 in QMdiArea::event(QEvent*) () from /usr/lib/libQt6Widgets.so.6
#18 0x00007ffff0103380 in QApplicationPrivate::notify_helper(QObject*, QEvent*) () from /usr/lib/libQt6Widgets.so.6
#19 0x00007ffff51b5d51 in Gui::GUIApplication::notify (this=0x7fffffffcc30, receiver=0x555555b08890, event=0x7fffffffc430) at /home/pieter/current/fpa/freecad-pr-21292/src/Gui/GuiApplication.cpp:103
#20 0x00007fffeef6bf48 in QCoreApplication::notifyInternal2(QObject*, QEvent*) () from /usr/lib/libQt6Core.so.6
#21 0x00007ffff015847d in QWidgetPrivate::hideChildren(bool) () from /usr/lib/libQt6Widgets.so.6
#22 0x00007ffff0158a4e in QWidgetPrivate::hide_helper() () from /usr/lib/libQt6Widgets.so.6
#23 0x00007ffff015af4a in QWidgetPrivate::setVisible(bool) () from /usr/lib/libQt6Widgets.so.6
#24 0x00007ffff0158cef in QWidgetPrivate::handleClose(QWidgetPrivate::CloseMode) () from /usr/lib/libQt6Widgets.so.6
#25 0x00007ffff017f492 in ?? () from /usr/lib/libQt6Widgets.so.6
#26 0x00007fffef815a3d in QWindow::event(QEvent*) () from /usr/lib/libQt6Gui.so.6
#27 0x00007ffff0103380 in QApplicationPrivate::notify_helper(QObject*, QEvent*) () from /usr/lib/libQt6Widgets.so.6
#28 0x00007ffff51b5d51 in Gui::GUIApplication::notify (this=0x7fffffffcc30, receiver=0x555555db2a30, event=0x7fffffffca60) at /home/pieter/current/fpa/freecad-pr-21292/src/Gui/GuiApplication.cpp:103
#29 0x00007fffeef6bf48 in QCoreApplication::notifyInternal2(QObject*, QEvent*) () from /usr/lib/libQt6Core.so.6
#30 0x00007fffef7a12a2 in QGuiApplicationPrivate::processCloseEvent(QWindowSystemInterfacePrivate::CloseEvent*) () from /usr/lib/libQt6Gui.so.6
#31 0x00007fffef7effc3 in QPlatformWindow::close() () from /usr/lib/libQt6Gui.so.6
#32 0x00007fffef812b2b in QWindow::close() () from /usr/lib/libQt6Gui.so.6
#33 0x00007ffff01475ce in QWidget::~QWidget() () from /usr/lib/libQt6Widgets.so.6
#34 0x00007ffff5c28044 in Gui::MainWindow::~MainWindow (this=0x7fffffffcc00) at /home/pieter/current/fpa/freecad-pr-21292/src/Gui/MainWindow.cpp:529
#35 0x00007ffff5077a5d in Gui::Application::runApplication () at /home/pieter/current/fpa/freecad-pr-21292/src/Gui/Application.cpp:2713
#36 0x000055555557cc4e in main (argc=4, argv=0x7fffffffd3e8) at /home/pieter/current/fpa/freecad-pr-21292/src/Main/MainGui.cpp:344
</pre>
</details>

I think that is unrelated to this PR, so I propose to ignore this for this PR.

pieterhijma replied:
It turns out that I don't have that issue on the main branch anymore. As this has recently be rebased on the main branch and includes many of the same commits, I think it's worth to investigate this a bit more.

tritao replied:
> It turns out that I don't have that issue on the main branch anymore. As this has recently be rebased on the main branch and includes many of the same commits, I think it's worth to investigate this a bit more.

I don't think it's related to these changes, but can you try this patch and see if it helps:

```diff
diff --git a/src/Gui/MainWindow.cpp b/src/Gui/MainWindow.cpp
index 4afcf1b644..XXXXXXXXXX 100644
--- a/src/Gui/MainWindow.cpp
+++ b/src/Gui/MainWindow.cpp
@@ -521,6 +521,13 @@ MainWindow::MainWindow(QWidget* parent, Qt::WindowFlags f)

MainWindow::~MainWindow()
{
+    // During QMainWindow/QWidget teardown Qt may still emit
+    // QMdiArea::subWindowActivated while child subwindows are being closed.
+    // Disconnect this callback before base-class destruction so it cannot
+    // re-enter MainWindow slots after the derived destructor has started.
+    if (d->mdiArea) {
+        disconnect(d->mdiArea, &QMdiArea::subWindowActivated, this, &MainWindow::onWindowActivated);
+    }
   delete d->status;
   delete d;
   instance = nullptr;
```

pieterhijma replied:
Great, thanks. That patch solves the issue!

pieterhijma reviewed (commented):
Ran the tests as well. Can't do the stress test.

tritao committed:
Gui: disconnect MDI activation callback during teardown

pieterhijma commented:
How do I test this now given all the other changes to partdesign? For example, polar pattern has now "recompute on change" that still blocks the GUI and computes intermediate results.

tritao commented:
> How do I test this now given all the other changes to partdesign? For example, polar pattern has now "recompute on change" that still blocks the GUI and computes intermediate results.

The idea has been to have this PR only add the base infrastructure, and for the next PR to adopt the async recompute feature on a bunch of features around FreeCAD. So I guess there's no obvious way to test UX responsiveness improvements strictly from this PR besides calling `Std_Refresh` command on a heavy document.

xtemp09 commented:
@tritao, `#ifdef`s at Application.cpp L 3500 and 3503 duplicate each other. Are `FC_OS_WIN32` and `FC_OS_MACOSX` not mutually exclusive?

tritao commented:
> @tritao, `#ifdef`s at Application.cpp L 3500 and 3503 duplicate each other. Are `FC_OS_WIN32` and `FC_OS_MACOSX` not mutually exclusive?

That just appears on the diff due to whitespace or re-formatting noise that crept up on the PR, the code itself is pre-existing code, but I think you are right and it should be cleaned up as separate commit.

tritao commented:
Updated again, addresses the latest feedback from @pieterhijma, including a stabilization fix for that unrelated crash, and cleans up a lot of formatting-only cruft from the changes.

pieterhijma commented on src/App/Document.cpp:
```diff
@@ -105,6 +107,17 @@ using namespace zipios;
+namespace
+{
+
+bool transactionStateBlocksRecoveryWrite(const DocumentP& documentPrivate)
```
Great, I think this is much better.

pieterhijma commented on src/App/RecoverySnapshot.cpp:
```diff
@@ -0,0 +1,178 @@
+// SPDX-License-Identifier: LGPL-2.1-or-later
+
+/***************************************************************************
+ *   Copyright (c) 2026 The FreeCAD project association AISBL              *
```
Not your own copyright? This is adapted from Gui/AutoSaver.cpp but heavily refactored. I checked whether it should be Werner's copyright, but I think it's safe to add your own copyright for this.

pieterhijma commented on src/App/RecoverySnapshot.h:
```diff
@@ -0,0 +1,44 @@
+// SPDX-License-Identifier: LGPL-2.1-or-later
+
+/***************************************************************************
+ *   Copyright (c) 2026 The FreeCAD project association AISBL              *
```
And this one as well.

pieterhijma reviewed (changes requested):
Almost done! Great work. Just a couple of tiny remarks.

Oh one other thing. `App/Document.cpp` has these strange ^M characters again, mixed mode. Could you please fix that as well? Thanks in advance!

tritao commented:
> Almost done! Great work. Just a couple of tiny remarks.
> 
> Oh one other thing. `App/Document.cpp` has these strange ^M characters again, mixed mode. Could you please fix that as well? Thanks in advance!

Thanks, please check the latest update and let me know if you'd like any additional changes.

tritao commented:
FYI CI failure is unrelated to the changes, just KDE Neon key network issues.. again.

pieterhijma reviewed (changes requested):
Unfortunately, this is not the solution. This change to Document associates all lines to you (git blame).

tritao commented:
> Unfortunately, this is not the solution. This change to Document associates all lines to you (git blame).

Yep my bad, just a CRLF vs. LF issue. Should be fixed now.

pieterhijma reviewed (approved):
Ok, thanks, looks good to me! Sorry for the back and forth, but I think it is important to get it right for this particular PR.

marioalexis84 commented:
@tritao auto-recovery file saving is triggered with every property change in App. This can take a long time depending on the objects in the document and significantly interferes with the program's workflow (try it with a dense FEM meshes or FEM pipelines).

tritao commented:
> @tritao auto-recovery file saving is triggered with every property change in App. This can take a long time depending on the objects in the document and significantly interferes with the program's workflow (try it with a dense FEM meshes or FEM pipelines).

Taking a look
